lib/spack/spack/solver/asp.py

```diff
@@ -1230,12 +1230,6 @@ def define_version_constraints(self):
                 if v.satisfies(versions)
             ]
 
-            # This is needed to account for a variable number of
-            # numbers e.g. if both 1.0 and 1.0.2 are possible versions
-            exact_match = [v for v in allowed_versions if v == versions]
-            if exact_match:
-                allowed_versions = exact_match
-
```

Comment:
Just removing this doesn't make any tests fail. Also worth noting that with this removed

```spack spec mpich@3.2```

Still concretizes to @3.2 exactly, it doesn't concretize to 3.2.1